components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/graaljs/JsGraalGraphBuilder.java

```diff
@@ -0,0 +1,707 @@
+            authenticationOptionsObj = IntStream.range(0, authenticationOptionsList.size())
+                    .boxed()
+                    .collect(Collectors.toMap(
+                            String::valueOf,
+                            authenticationOptionsList::get
+                                             ));
+        }
+
+        if (authenticationOptionsObj instanceof Map) {
+            filterOptions((Map<String, Map<String, String>>) authenticationOptionsObj, stepConfig);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Authenticator options not provided or invalid, hence proceeding without filtering");
+            }
+        }
+
+        Object authenticatorParams = options.get(FrameworkConstants.JSAttributes.AUTHENTICATOR_PARAMS);
+        if (authenticatorParams instanceof Map) {
+            authenticatorParamsOptions((Map<String, Object>) authenticatorParams, stepConfig);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Authenticator params not provided or invalid, hence proceeding without setting params");
+            }
+        }
+
+        Object stepOptions = options.get(FrameworkConstants.JSAttributes.STEP_OPTIONS);
+        if (stepOptions instanceof Map) {
+            handleStepOptions(stepConfig, (Map<String, String>) stepOptions, stepConfigMap);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Step options not provided or invalid, hence proceeding without handling");
+            }
+        }
+    }
+
+    public static void sendErrorAsync(String url, Map<String, Object> parameterMap) {
+
+        FailNode newNode = createFailNode(url, parameterMap, true);
+
+        AuthGraphNode currentNode = dynamicallyBuiltBaseNode.get();
+        if (currentNode == null) {
+            dynamicallyBuiltBaseNode.set(newNode);
+        } else {
+            attachToLeaf(currentNode, newNode);
+        }
+    }
+
+    private static FailNode createFailNode(String url, Map<String, Object> parameterMap, boolean isShowErrorPage) {
+
+        FailNode failNode = new FailNode();
+        if (isShowErrorPage && StringUtils.isNotBlank(url)) {
+            failNode.setErrorPageUri(url);
+        }
+        // setShowErrorPage is set to true as sendError function redirects to a specific error page.
+        failNode.setShowErrorPage(isShowErrorPage);
+
+        parameterMap.forEach((key, value) -> failNode.getFailureData().put(key, String.valueOf(value)));
+        return failNode;
+    }
+
+    /**
+     * Javascript based Decision Evaluator implementation.
+     * This is used to create the Authentication Graph structure dynamically on the fly while the authentication flow
+     * is happening.
+     * The graph is re-organized based on last execution of the decision.
+     */
+    public class JsBasedEvaluator implements AuthenticationDecisionEvaluator {
```

Comment:
check if this needs to be a public class

Reply:
This will be using thread locals from the Graph builder. Hence it needs to be defined as an inner class. Further, it needs to be a public class since it implements a public interface which is used in other classes.